### DISCUSSION TITLE
Self-Signed from NginxProxyManager to Zeroxy, Problem with Certificate

### DISCUSSION THREAD

**Discussion by Chiller2019** [tobychui/zoraxy#371] [Q&A] [answered]
Hey,

I have a question.
With NginxProxyManager you can include a wildcard certificate and use it for each host. For this I have packed each container including NginxProxyManager (or now Zeroxy) in its own network.

I didn't open any ports at the time because Nginx didn't need them.

With Zeroxy, I integrated my wildcard certificate as TLS and as fallback TLS. However, it cannot open the fixed IP of the container even though the IP and port are specified.

Does anyone here have experience with self-signed and Zeroxy to explain this to me quickly?

Kind regards

//When I try to ping the server via Network Tools -> Traceroute/Ping, TCP connection refused is returned although they are in the same network

**Comment by tobychui** [comment]
Hey @Chiller2019 ,

Can you show a screenshot of your proxy rules and setup here? It seems like a networking issue to me than a bug in Zoraxy.

**Reply by Chiller2019** [reply]
Sure, one Minute.

Problem is, why did it run without problems with the NginxProxyManager? - The IPs could be found, the IP scanner also shows the IPs, as an example.

**Reply by Chiller2019** [reply]
IP-Scanner:
![grafik](https://github.com/user-attachments/assets/e449314b-cfd5-4344-b0a6-41b3f3a4fd22)
Old Config of NGinxProxyManager (Custom is the Wildcard-Certificate because its self-signed and not with ACME, dont use ACME):
![grafik](https://github.com/user-attachments/assets/bad461ac-24a8-48f7-949e-688caaa4c409)
Zoraxy HTTP-Proxy:
![grafik](https://github.com/user-attachments/assets/61626392-3249-46a4-972f-9346bf2ca477)
TLS:
![grafik](https://github.com/user-attachments/assets/1a42622a-2b17-45c8-8378-ee5466f20d94)
Portainer-Network (all with fixed-ips):
![grafik](https://github.com/user-attachments/assets/5e482e25-0320-4259-a980-0d007b10823f)
Portainer-Container-PDF:
![grafik](https://github.com/user-attachments/assets/2a94f5fc-cb98-4383-96c9-ce0015aa737c)
Portainer-Container-Zoraxy:
![grafik](https://github.com/user-attachments/assets/ab4a45ab-8c92-438f-adc1-afc3f57723db)
Enviroments of Zoraxy:
![grafik](https://github.com/user-attachments/assets/a269a292-12a9-40d6-99d4-ebfa1aed656f)
Zoraxy-Global settings:
![grafik](https://github.com/user-attachments/assets/49a81fb4-606e-42b3-ac49-1ac8c0bd1202)

**Reply by tobychui** [reply]
The settings all seems ok to me, have you tried running Zoraxy without docker (e.g. run it on your PC locally, with the same setup) and see if it works? 

And your PDF container have no published port in your screenshot, you sure that is correct?

**Reply by Chiller2019** [reply]
That's what I thought.
I haven't tried that yet either, but it's also more difficult because of the effort involved. These are split into a separate network and can only be accessed via the proxy, you shouldn't be able to access them easily via IP or port but only via the preset domain.

I would be interested to know whether a proxy rule really uses this characteristic. Although it should still be able to access it via docker.sock.

And yes, the container does not need a public port. NxingProxyManager did not need that. Not sure why either.
But I'll try to open the port, it only works as a user of the “StaticNetwork” anyway

**Reply by Chiller2019** [reply, marked as answer]
Ok, the error is fixed. Zoraxy had to be restarted.
~~The question is whether I have to restart Zoraxy every time I start a new service or not.~~

Port does not have to be set, it worked the same way for me

**Reply by Chiller2019** [reply]
@tobychui Thx for your quick answer!

**Comment by Trackhe** [comment]
How did you upload your own wildcard cert because i see only the option where i can upload the private and public key.
If i upload them i get only the error: ERR_SSL_UNRECOGNIZED_NAME_ALERT in browser

**Reply by tobychui** [reply]
@Trackhe when you say you own a wildcard cert but not request from, let say Lets' Encrypt, I assume you have purchased it from a CA somewhere. You should contact the seller and ask them instead.

**Reply by Trackhe** [reply]
I have my own CA. I only don't understand how i can upload it correctly because the ui says upload private key and public key but no certificat no intermediate certificate as NPM does. There it is so easy to do that. But in zoraxy i don't see where i can do that exept the option with uploading the keys but that dosn't work

**Reply by tobychui** [reply]
Hmm I can't read German but I guess you mean intermediate certificate? There is a build in tool in Zoraxy where you can convert the certificate into Zoraxy supported format, I remember there is a link to it in the TLS / SSL page.

**Reply by tobychui** [reply]
![圖片](https://github.com/user-attachments/assets/75a93d94-cf73-4857-b1d7-0eabedaec3c8)
Oh it is here, are you looking for this?

**Reply by Trackhe** [reply]
> ![圖片](https://private-user-images.githubusercontent.com/24617523/399950360-75a93d94-cf73-4857-b1d7-0eabedaec3c8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU5MTA1NzUsIm5iZiI6MTczNTkxMDI3NSwicGF0aCI6Ii8yNDYxNzUyMy8zOTk5NTAzNjAtNzVhOTNkOTQtY2Y3My00ODU3LWIxZDctMGVhYmVkYWVjM2M4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTAzVDEzMTc1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU0MjBlZTUwNDIzOTAzNDkzMjViNzlhNzljM2QxNmU3ODZlNjg1NzViMDgyMzJiMzMxYWUyZGFlYzczMjQ0YTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Cn6xOJFYvbeIlQUqi-gHT1kwrXNwKIysDEGMbWRiMeE) Oh it is here, are you looking for this?

I tried that also, i can upload my ca, the intermediate certificate and get a .pem file back. but how i can use it then in combination with my certificate that includes the commonName like *.example.com?

<img width="1264" alt="Bildschirmfoto 2025-01-03 um 14 21 28" src="https://github.com/user-attachments/assets/4a090633-0b6c-42bf-b76c-87876b8d83f0" />

There are only upload field for private key and public key

srry for my typos btw. and thank you for your support!

**Reply by tobychui** [reply]
Common Name? It has been [deprecated since 2011 for the purpose of verification](https://github.com/caddyserver/caddy/issues/3755).
Now the CN field only serve as inspection purposes.

> Common Name attribute is only deprecated for the purpose of certificate verification, the attribute still serves a valid non-deprecated purpose for administration/inspection.

I guess in Zoraxy you can still "force serve" it by naming it the longest common suffix as the Server Name (e.g. fill in "example.com" in the Server Name field, instead of *.example.com).

**Reply by Trackhe** [reply]
I use also SAN(i have filled the CN and SAN field with *.example.com(mydomain oc)) but that is not the point, the point is i don't know how to upload the certificate / pam which i generate with the Open Conversion Tool and how to upload the site certificate

and in Server Name i put example.com without *. or _. in because it says in the TIPs and SAN section

"If you have a wildcard certificate that covers *.example.com, you can just enter example.com as server name to add a certificate."

but i even tried it with and without, all combinations ^^ but nothing works. It seems that zoraxy try to provide the right certificate but when i only upload the keys there is no certificate to provide. And as i said i don't know how to upload the certificate because i only see there are upload buttons for the keys

**Reply by tobychui** [reply]
I have no idea what you are talking about. Even if in worst case Zoraxy cannot find a matching certificate via its internal SNI, you should see the Zoraxy self-signed certificate instead of, let say SSL error with "no certificate to provide" as quoted. Are you sure you are actually accessing / reaching your Zoraxy instance?

FYI, here is the cert you shd see if you have no certificate configured / no matching SAN certificate found

![圖片](https://github.com/user-attachments/assets/5cd34e64-5853-4b83-90c6-f3694dc3ca67)

**Reply by Trackhe** [reply]
Yes, if i don't provide any matching server name, and no custom keys Zoraxy provide the self-host cert. That works. but i don't know how to upload any custom certificate because there are only keys uploads no cert uploads on the cert settings page.

**Reply by tobychui** [reply]
Hmm, not sure if this helps? [https://github.com/tobychui/zoraxy/issues/10#issuecomment-1566656725](https://github.com/tobychui/zoraxy/issues/10#issuecomment-1566656725)

**Reply by Trackhe** [reply]
That was not really the solution because i followed his instruction exactly and it wasn't really working. But it was motivating me to tinker a bit with different combinations and in combination with the error messages in the container log i got a combination working.

On the "public key field" i uploaded 

domain.crt
```
-----BEGIN PUBLIC KEY-----
public key
-----END PUBLIC KEY-----
-----BEGIN CERTIFICATE-----
cert
-----END CERTIFICATE-----
-----BEGIN CERTIFICATE-----
intermediate cert
-----END CERTIFICATE-----
```

and on the "private key field"
```
-----BEGIN RSA PRIVATE KEY-----
private key
-----END RSA PRIVATE KEY-----
```

that seems working.


@tobychui maybe you can rename 

<img width="403" alt="Bildschirmfoto 2025-01-03 um 15 45 04" src="https://github.com/user-attachments/assets/fb33d44c-4378-4542-96a2-db4e48a70f08" />

to something like "Public Key + Certificate + Intermediate Certificate (optional)"

**Reply by tobychui** [reply]
Thanks for the idea. I think in cases like this, a better UI will be needed to minimize confusion. I will think about it later this weekend.
